eth/vm/computation.py

```diff
@@ -546,25 +565,51 @@ def apply_computation(cls,
         Perform the computation that would be triggered by the VM message.
         """
         with cls(state, message, transaction_context) as computation:
-            # Early exit on pre-compiles
-            if message.code_address in computation.precompiles:
-                computation.precompiles[message.code_address](computation)
-                return computation
+            if computation.state.tracer is None:
+                # Early exit on pre-compiles
+                if message.code_address in computation.precompiles:
+                    computation.precompiles[message.code_address](computation)
+                    return computation
 
             for opcode in computation.code:
                 opcode_fn = computation.get_opcode_fn(opcode)
+                pc = max(0, computation.code.pc - 1)
 
                 computation.logger.trace(
                     "OPCODE: 0x%x (%s) | pc: %s",
                     opcode,
                     opcode_fn.mnemonic,
-                    max(0, computation.code.pc - 1),
+                    pc,
                 )
 
+                if computation.state.tracer is not None:
+                    """
+                    TODO: once we are able to predict gas_used, we can capture_state here,
+                          no dumping of stack and memory required
+                    """
+                    gas_remaining_prev = computation._gas_meter.gas_remaining
+                    stack_prev = deepcopy(computation._stack)
+                    memory_prev = deepcopy(computation._memory)
```

Comment:
Once we can predict gas usage beforehand, we can avoid storing computation state before execution.